src/libmw/src/models/tx/TxBody.cpp

```diff
@@ -0,0 +1,210 @@
+{
+    return std::accumulate(
+        m_kernels.cbegin(), m_kernels.cend(), (uint64_t)0,
+        [](const uint64_t sum, const auto& kernel) noexcept { return sum + kernel.GetPegIn(); }
+    );
+}
+
+std::vector<PegOutCoin> TxBody::GetPegOuts() const noexcept
+{
+    std::vector<PegOutCoin> pegouts;
+    for (const Kernel& kernel : m_kernels) {
+        if (kernel.HasPegOut()) {
+            pegouts.push_back(kernel.GetPegOut().value());
+        }
+    }
+    return pegouts;
+}
+
+uint64_t TxBody::GetTotalFee() const noexcept
+{
+    return std::accumulate(
+        m_kernels.cbegin(), m_kernels.cend(), (uint64_t)0,
+        [](const uint64_t sum, const auto& kernel) noexcept { return sum + kernel.GetFee(); }
+    );
+}
+
+int64_t TxBody::GetSupplyChange() const noexcept
+{
+    int64_t coins_added = 0;
+    for (const Kernel& kernel : m_kernels) {
+        coins_added += kernel.GetSupplyChange();
+    }
+
+    return coins_added;
+}
+
+Serializer& TxBody::Serialize(Serializer& serializer) const noexcept
+{
+    return serializer
+        .AppendVec<Input>(m_inputs)
+        .AppendVec<Output>(m_outputs)
+        .AppendVec<Kernel>(m_kernels)
+        .AppendVec<SignedMessage>(m_ownerSigs);
+}
+
+TxBody TxBody::Deserialize(Deserializer& deserializer)
+{
+    std::vector<Input> inputs = deserializer.ReadVec<Input>();
+    std::vector<Output> outputs = deserializer.ReadVec<Output>();
+    std::vector<Kernel> kernels = deserializer.ReadVec<Kernel>();
+    std::vector<SignedMessage> owner_sigs = deserializer.ReadVec<SignedMessage>();
+    return TxBody(std::move(inputs), std::move(outputs), std::move(kernels), std::move(owner_sigs));
+}
+
+void TxBody::Validate() const
+{
+    // Verify weight
+    if (Weight::ExceedsMaximum(*this)) {
+        ThrowValidation(EConsensusError::BLOCK_WEIGHT);
+    }
+
+    // Verify no kernel's extra_data or pegout scriptPubKey exceeds max size
+    bool extra_data_exceeds_max = std::any_of(
+        m_kernels.cbegin(), m_kernels.cend(),
+        [](const Kernel& kernel) {
+            size_t pubkeySize = kernel.GetPegOut().has_value() ? kernel.GetPegOut().value().GetScriptPubKey().size() : 4;
+            return pubkeySize > 42 || pubkeySize < 4 || kernel.GetExtraData().size() > libmw::MAX_KERNEL_EXTRADATA_SIZE;
+        }
+    );
+    if (extra_data_exceeds_max) {
+        ThrowValidation(EConsensusError::BLOCK_WEIGHT);
+    }
+
+    // Verify inputs, outputs, kernels, and owner signatures are sorted
+    if (!std::is_sorted(m_inputs.cbegin(), m_inputs.cend(), SortByCommitment)
+        || !std::is_sorted(m_outputs.cbegin(), m_outputs.cend(), SortByCommitment)
+        || !std::is_sorted(m_kernels.cbegin(), m_kernels.cend(), KernelSort)
+        || !std::is_sorted(m_ownerSigs.cbegin(), m_ownerSigs.cend(), SortByHash))
+    {
+        ThrowValidation(EConsensusError::NOT_SORTED);
+    }
+
+    // Verify no duplicate inputs
+    std::unordered_set<Commitment> input_commits;
+    std::transform(
```

Comment:
Would be nice to factor this out into a function as its repeated thrice

Reply:
Maybe using Commitments::From although that forms a vector rather than a set

Reply:
Created a Commitments::SetFrom